### PR TITLE
TransactionMarker middleware

### PR DESCRIPTION
**Pull request by craig410**

### CODE REVIEW

**Comment by craig410** [comment]
@acoulton pending release of https://github.com/ingenerator/php-utils/pull/38 first

I'll have a shot at sorting factory method for adding to the end of the default chain tomorrow - forgot about it

File: test/unit/Server/Middleware/TransactionMarkerMiddlewareTest.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            ['headers' => ['X-Transaction' => 'abcd-1234', 'X-Transaction-Expire' => '2021-02-04 01:02:03']]
+        );
+
+        $this->assertSame(
+            [
+                'code'        => TransactionMarkerResult::TRANSACTION_EXPIRED,
+                'msg'         => 'transaction rolled back',
+                'log_context' => [
+                    'uuid'   => 'abcd-1234',
+                    'expiry' => '2021-02-04 01:02:03',
```
Comment (acoulton) [inline test/unit/Server/Middleware/TransactionMarkerMiddlewareTest.php]: Minor stylistic thing, I prefer to consistently use either stubbed time or real time. E.g. if the class is using RealtimeClock then use fixed time strings for "now" and the test value. 

But in this case as the class under test is using `new DatetimeImmutable` then the test value should be eg `Datestring::ymdhis(new DatetimeImmutable('-1s'))`.

Not really any way it can make a difference with something this simple but it helps make sure we're testing boundary conditions and avoids risk of oddities.

File: src/Server/Middleware/TransactionMarkerMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        $expiry = DateTimeImmutableFactory::fromYmdHisInput(
+            $request->getHttpRequest()->getHeaderLine('X-Transaction-Expire')
+        );
+
+        // try again later, we're not ready yet
+        if (new DateTimeImmutable() < $expiry) {
+            return TransactionMarkerResult::notYetVisible($uuid, $expiry);
+        }
+
+        // quit retrying / it's been completed elsewhere
```
Comment (acoulton) [inline src/Server/Middleware/TransactionMarkerMiddleware.php]: `quit retrying - most likely transaction never completed e.g. rolled back` is maybe a better explanation of what's happening here?

**Review by acoulton — APPROVED** [review]
Awesome work. 2 tiny nit-picks to show I read it properly 🤣

**Comment by acoulton** [comment]
> I'll have a shot at sorting factory method for adding to the end of the default chain tomorrow - forgot about it

Good point. Also I think we should maybe add the two new result codes to the `TaskResultCodeMapper` just so we ship with defaults? Prob with the loglevels dialled down to notice/warning for ongoing use, we can always bump them up at project level to keep an eye when we first ship.

File: test/integration/TaskServerFactoryTest.php [hunk 3]
```diff
@@ -23,7 +25,8 @@ public function test_can_create_server_stack()
             $this->createStub(CacheItemPoolInterface::class),
             $this->createStub(TaskHandlerFactory::class),
             [],
-            []
+            [],
+            $this->createStub(TransactionMarkerMiddleware::class)
```
Comment (acoulton) [inline test/integration/TaskServerFactoryTest.php]: I guess ideally we'd assert this middleware is actually registered (and at the end), tho it's not going to be that easy to do it.... Maybe just have to use ObjectPropertyRipper to inspect the internals in this case. Wasn't a concern before because all the other deps are required.

In which case possibly also worth a test with two extra middlewares to make sure the variadic doesn't get lost along the way in future.

**Review by acoulton — COMMENTED** [review]
Awesome.
